src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs

```diff
@@ -133,9 +134,9 @@ private void Add(string path, object value, T objectToApplyTo, Operation<T> oper
                 // what if it's an array but there's no position??
                 if (IsNonStringArray(patchProperty))
                 {
-                    // now, get the generic type of the enumerable
+                    // now, get the generic type of the enumerable from the runtime type of patchProperty value.
                     var genericTypeOfArray = PropertyHelpers.GetEnumerableType(
-                        patchProperty.Property.PropertyType);
+                        patchProperty.Property.ValueProvider.GetValue(patchProperty.Parent).GetType());
```

Comment:
`JsonProperty`'s `PropertyType` returns declared type of the property not the runtime type. So if there is a property of declared type `IList`, this call returns `null`. So I had to change it to take the  runtime type of value.

Its a good point, I have to handle the case where runtime type value is `null`.